### DISCUSSION TITLE
Why is ReadOnlySpan<char> overload is chosen over string when inheritance is involved? even when the argument is string?

### DISCUSSION THREAD

**Discussion by fitdev** [dotnet/csharplang#6316] [Q&A] [answered]
Consider [this code](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgYhgVwDZ4TB4CmABCTEaQLABQ9AAgEwCMTAzGS2QMJkBvemRHcuANygAnBDgh4ywAPZKFAWQAUjVgAYA/GVgAHHAgCUg4aOuMA7GQRScJANxWRAX3peGdRlyIURwgAYwRuZjIAETIQPks6Gy4lcRIpKSgAE3JlVTJNbX1DGBNzBOsbe2BUEgA6TWNTMzdE0R9rdzEyQOCwxRV1DQAlEghMgHkYPABPAGUjCBgAHhCACwgpAD5i0ubvek5uVgA2CLIAFRIg8s7/I9PGABYyIZwYDRjMiyFW63ENhxXBCFMgAXjIACIAEKEVYQloVMiZeoaBBAwrNESddqibxAA=):

```cs
public class C {
    public virtual bool M(string? input) {
        return true;
    }
}

public abstract class D : C {
    public override bool M(string? input) {
        return base.M(input);
    }
    
    public abstract bool M(ReadOnlySpan<char> input);
}

public static class Test {
 
    public static void Run(D d) {
      var teststring = "Blah";
      d.M(teststring);  
    }
    
}
```

Question: why in the `Test.Run` routine `D.M(ReadOnlySpan<char> input)` is chosen over `D.M(string? input)` even though the argument passed is explicitly typed as `string`, i.e. `var teststring = "Blah";`

**Comment by svick** [comment, marked as answer]
> why in the `Test.Run` routine `D.M(ReadOnlySpan<char> input)` is chosen over `D.M(string? input)`

What is happening here is that `D.M(ReadOnlySpan<char>)` is chosen over <code>**C**.M(string?)</code>, because methods from derived classes are considered before methods from base classes. And, for the purposes of overload resolution, `D.M(string?)` does not exist.

**Reply by fitdev** [reply]
Thank you for a quick reply. I figured it may have something to do with inheritance. Is there a way to explicitly call the method I want without resorting to renaming it?

**Reply by PathogenDavid** [reply]
If you cast `d` to `C` it will call `M(string)`. IE: `((C)d).M(teststring);`
